feat: Flutter v0-v1 upgrade guide (#5358)
aws-amplify/docs: src/fragments/lib/project-setup/flutter/upgrade-guide/upgrade-guide.mdx

> With the release of v1 amplify-flutter now supports Web and Desktop for Auth, API, Analytics, and Storage use cases. Developers can now build cross-platform Flutter apps with Amplify that target iOS, Android, Web, and Desktop (macOS, Windows, Linux) using a single codebase.

Reviewer: ```suggestion
With the release of v1, Amplify Flutter now supports Web and Desktop for Auth, API, Analytics, and Storage use cases. Developers can build cross-platform Flutter apps with Amplify that target iOS, Android, Web, and Desktop (macOS, Windows, Linux) using a single codebase.
```

We have re-written our libraries in Dart. In some places, we have made breaking changes to improve ergonomics or enable features that had been missing from the v0 implementations.

## Prerequisites

- A Flutter application targeting Flutter SDK >= 3.3.0 with Amplify libraries integrated

> Amplify requires a minimum target platform for iOS (13.0), Android (API level 24), and macOS (10.15). Additional setup is required for some target platforms. Please see the [platform setup](/lib/project-setup/platform-setup) guide for more details on platform specific setup.

Reviewer: ```suggestion
Amplify requires a minimum target platform for iOS (13.0), Android (API level 24), and macOS (10.15). Additional setup is required for some target platforms. Please see the [platform setup](/lib/project-setup/platform-setup) guide for more details on platform-specific setup.
```

## Auth

> - `fetchAuthSession` had several changes. See the example below for how to migrate.

Reviewer: ```suggestion
- `fetchAuthSession` has several changes. See the example below for how to migrate.
```

>   - `Amplify.Auth.getPlugin()` has been added to return plugin specific sub types. Casting to `CognitoAuthSession` is no longer needed when using this.

Reviewer: ```suggestion
  - `Amplify.Auth.getPlugin()` has been added to return plugin-specific subtypes. Casting to `CognitoAuthSession` is no longer needed when using this method.
```

>   - `getAWSCredentials` has been deprecated. AWS credentials will always be included in the response.

Reviewer: ```suggestion
  - The `getAWSCredentials` flag was previously needed to retrieve AWS credentials. This flag has been deprecated and is no longer needed. AWS credentials will always be included in the response.
```

  - `fetchAuthSession` will no longer throw exceptions. Instead, individual getters of type `AuthResult` will throw if the underlying operation fails.
    - _Why?_ The previous behavior did not allow for partial successes in offline scenarios when user pool tokens were still valid, but AWS credentials were expired. See [amplify-flutter#760](https://github.com/aws-amplify/amplify-flutter/issues/760#issuecomment-1453559459) for more details.

>   - `CognitoAuthSession.credentials`, `CognitoAuthSession.userPoolTokens`, `CognitoAuthSession.userSub`, and `CognitoAuthSession.identityId` are deprecated and will be removed in a future release.

Reviewer: ```suggestion
  - `CognitoAuthSession.credentials`, `CognitoAuthSession.userPoolTokens`, `CognitoAuthSession.userSub`, and `CognitoAuthSession.identityId` are deprecated and will be removed in a future release. Use `CognitoAuthSession.credentialsResult`, `CognitoAuthSession.userPoolTokensResult`, `CognitoAuthSession.userSubResult`, and `CognitoAuthSession.identityIdResult` instead.
```

  // the users session has expired.
} on NetworkException {
  // the access and/or id token is expired but cannot be refreshed because the
  // user is offline.
}

// `AuthResult.valueOrNull` will return the identityId if present or null,
// but will never throw even if the underlying operation failed.
final identityId = session.identityIdResult.valueOrNull;
safePrint('identityId: $identityId');
```

- The next step returned from auth operations are now enums. See the example below for how to migrate for `Auth.signUp()`.

```dart
// v0
final result = await Amplify.Auth.signUp(/* ... */);
switch (result.nextStep?.signUpStep) {
  case 'CONFIRM_SIGN_UP':
    // prompt user to confirm sign in using nextStep.codeDeliveryDetails
    break;
  case 'DONE':
    // sign up is complete, user can be taken to an authenticated state.
    break;
}
```

```dart
// v1
final result = await Amplify.Auth.signUp(/* ... */);
switch (result.nextStep.signUpStep) {
  case AuthSignUpStep.confirmSignUp:
    // prompt user to confirm sign in using nextStep.codeDeliveryDetails
    break;
  case AuthSignUpStep.done:
    // sign up is complete, user can be taken to an authenticated state.
    break;
}
```

> - Auth hub events are now have a specific sub type, `AuthHubEvent` and contain a `type` that is an enum. See the example below for how to migrate.

Reviewer: ```suggestion
- Auth hub events now have a specific subtype (`AuthHubEvent`) and contain a `type` that is an enum. See the example below for how to migrate.
```

```dart
// v0
final hubSubscription = Amplify.Hub.listen([HubChannel.Auth], (hubEvent) {
  switch(hubEvent.eventName) {
    case 'SIGNED_IN':
      safePrint('User is signed in.');
      break;
    case 'SIGNED_OUT':
      safePrint('User is signed out.');
      break;
    case 'SESSION_EXPIRED':
      safePrint('The session has expired.');
      break;
    case 'USER_DELETED':
      safePrint('The user has been deleted.');
      break;
  }
});
```

```dart
// v1
final subscription = Amplify.Hub.listen(HubChannel.Auth, (AuthHubEvent event) {
  switch (event.type) {
    case AuthHubEventType.signedIn:
      safePrint('User is signed in.');
      break;
    case AuthHubEventType.signedOut:
      safePrint('User is signed out.');
      break;
    case AuthHubEventType.sessionExpired:
      safePrint('The session has expired.');
      break;
    case AuthHubEventType.userDeleted:
      safePrint('The user has been deleted.');
      break;
  }
});
```

> - Cognito specific API options such as `CognitoSignInOptions` are deprecated in favor of `pluginOptions`. The classes will be removed in a future release. See below for an example of how to migrate for `Auth.signIn()`.

Reviewer: ```suggestion
- Cognito-specific API options such as `CognitoSignInOptions` are deprecated in favor of `pluginOptions`. The classes will be removed in a future release. See below for an example of how to migrate for `Auth.signIn()`.
```

```dart
// v0
final result = await Amplify.Auth.signIn(
  username: username,
  password: password,
  options: const CognitoSignInOptions(
    clientMetadata: {'data': 'value'},
  ),
);
```

```dart
// v1
final result = await Amplify.Auth.signIn(
  username: username,
  password: password,
  options: const SignInOptions(
    pluginOptions: CognitoSignInPluginOptions(
      clientMetadata: {'data': 'value'},
    ),
  ),
);
```

## Analytics

- `AnalyticsProperties` renamed to `CustomProperties`
- `AnalyticsUserProfile` renamed to `UserProfile`
- `AnalyticsUserProfileLocation` renamed to `UserProfileLocation`
- In v0, the `autoFlushEventsInterval` in `amplifyconfiguration.dart` was read as seconds on iOS and milliseconds on Android. v1 fixes this behavior by reading `autoFlushEventsInterval` as seconds across all platforms.

> - in v1 flushing occurs at 30 second intervals. In v0 default auto flush iOS was 60 seconds, Android was 30 seconds.

Reviewer: ```suggestion
- In v1, auto-flushing of events occurs at 30 second intervals. In v0, the default auto flush iOS was 60 seconds and Android was 30 seconds.
```

> - in v1 session pause and resume are not automatically tracked. In v0 iOS auto tracked session pause and resume events.

Reviewer: ```suggestion
- In v1, session pause and resume events are not automatically tracked. In v0, iOS had auto-tracked session pause and resume events.
```

Reviewer: What does this mean, though? As a customer, I would wonder how to enable it or if I need it.

Author: @fjnoyp ?

Reviewer 2: One sec, looking at this

Reviewer 2: - In v1, to ensure consistent platform behavior, backgrounding and foregrounding the app will track and trigger a new session.  This is different than v0's behavior on iOS, which tracks session Pause and Resume events on app background/foreground.

Reviewer 2: Let's use Jordan's suggested rephrasing for this:

There was a platform inconsistency in v0. iOS paused the session when the app went into background, whereas Android stopped it and then started a new session on app resume. In v1, the behavior will match android from v0.

Reviewer 2: v0 behavior: iOS tracked session resume/pause. iOS and Android tracked session start/stop events. 
v1 behavior: All platforms track session resume/pause.

Reviewer 2: Apologies for the delay here.  Thinking this rephrase could be good: 

Automatic session reporting behavior has been aligned in v1.  For all platforms, backgrounding and foregrounding the app will stop and start a new session.  Previously on iOS, the session would pause and then resume.

> - You can now save to the `UserAttributes` field of a Pinpoint Endpoint by sending a `AWSPinpointUserProfile` to `identifyUser`.

Reviewer: ```suggestion
- You can now save to the `UserAttributes` field of a Pinpoint Endpoint by passing an `AWSPinpointUserProfile` to `identifyUser`.
```

      safePrint('User is signed in.');
      break;
    case AuthHubEventType.signedOut:
      safePrint('User is signed out.');
      break;
    case AuthHubEventType.sessionExpired:
      safePrint('The session has expired.');
      break;
    case AuthHubEventType.userDeleted:
      safePrint('The user has been deleted.');
      break;
  }
});
```

- Cognito specific API options such as `CognitoSignInOptions` are deprecated in favor of `pluginOptions`. The classes will be removed in a future release. See below for an example of how to migrate for `Auth.signIn()`.

```dart
// v0
final result = await Amplify.Auth.signIn(
  username: username,
  password: password,
  options: const CognitoSignInOptions(
    clientMetadata: {'data': 'value'},
  ),
);
```

```dart
// v1
final result = await Amplify.Auth.signIn(
  username: username,
  password: password,
  options: const SignInOptions(
    pluginOptions: CognitoSignInPluginOptions(
      clientMetadata: {'data': 'value'},
    ),
  ),
);
```

## Analytics

- `AnalyticsProperties` renamed to `CustomProperties` 
- `AnalyticsUserProfile` renamed to `UserProfile`
- `AnalyticsUserProfileLocation` renamed to `UserProfileLocation`

> - In v1 you can now specify autoFlushEventsInterval in seconds in amplifyConfiguration.dart to change how often cached events are flushed. v0 does not support this due to differences in iOS and Android implementations.

Reviewer 2: ```suggestion
- In v0, the autoFlushEventsInterval in amplifyconfiguration.dart was read as seconds on iOS and milliseconds on Android. v1 fixes this behavior by reading autoFlushEventsInterval as seconds across all platforms.
```

## Prerequisites

* A Flutter application targeting Flutter SDK >= 3.0.0 with Amplify libraries integrated
  
  The following are also required, depending on which platforms you are targeting:

  * An iOS configuration targeting at least iOS 13.0 and XCode version >=13.2
  * An Android configuration targeting at least Android API level 24 (Android 7.0) or above
  * Any browser supported by Flutter for Web (you can check the list of supported browsers [here](https://docs.flutter.dev/development/platform-integration/web/faq#which-web-browsers-are-supported-by-flutter))
  * Any Windows OS meeting Flutter minimums
  * macOS version 10.15 or higher
  * Any Ubuntu Linux distribution meeting Flutter minimums
  * For a full example please follow the [project setup walkthrough](/lib/project-setup/create-application)

## Auth

- Changes to API options
- Changes to exception handling
- Changes to nextStep types
- getPlugin for Cognito-specific behavior
- Changes to Hub events

## Analytics

- `AnalyticsProperties` renamed to `CustomProperties` 
- `AnalyticsUserProfile` renamed to `UserProfile`
- `AnalyticsUserProfileLocation` renamed to `UserProfileLocation`

> - In amplifyConfiguration.dart, the docs previously said you could configure autoFlushEventsInterval in milliseconds. Now, you can set autoFlushEventsInterval in seconds.

Reviewer 2: I think we can rewrite this to: 

- In v1 you can now specify `autoFlushEventsInterval` in seconds in `amplifyConfiguration.dart` to change how often cached events are flushed.   v0 does not support this due to differences in iOS and Android implementations.

- Previously Analytics was built on top of existing Amplify Android and iOS functionality. Those platforms behaved differently but now behave the same:
    - in v1 flushing occurs at 30 second intervals. In v0 default auto flush iOS was 60 seconds, Android was 30 seconds.
    - in v1 session pause and resume are not automatically tracked. In v0 iOS auto tracked session pause and resume events. 
- You can now save to the `UserAttributes` field of a Pinpoint Endpoint by sending a `AWSPinpointUserProfile` to `identifyUser`.

> - Analytics Endpoint metadata is stored using in a different namespace/location.  Existing metadata will automatically be migrated.

Reviewer 2: ```suggestion
- Analytics Endpoint metadata is stored in a different namespace/location.  Existing metadata will automatically be migrated.
```

With the release of v1 amplify-flutter now supports Web and Desktop for Auth, API, Analytics, and Storage use cases.  Developers can now build cross-platform Flutter apps with Amplify that target iOS, Android, Web, and Desktop (macOS, Windows, Linux) using a single codebase.

We have re-written our libraries in Dart. In some places, we have made breaking changes to improve ergonomics or enable features that had been missing from the v0 implementations.

## Prerequisites

> * A Flutter application targeting Flutter SDK >= 3.0.0 with Amplify libraries integrated

Reviewer 3: ```suggestion
* A Flutter application targeting Flutter SDK >= 3.3.0 with Amplify libraries integrated
```

>   The following are also required, depending on which platforms you are targeting:
> 
>   * An iOS configuration targeting at least iOS 13.0 and XCode version >=13.2
>   * An Android configuration targeting at least Android API level 24 (Android 7.0) or above
>   * Any browser supported by Flutter for Web (you can check the list of supported browsers [here](https://docs.flutter.dev/development/platform-integration/web/faq#which-web-browsers-are-supported-by-flutter))
>   * Any Windows OS meeting Flutter minimums
>   * macOS version 10.15 or higher
>   * Any Ubuntu Linux distribution meeting Flutter minimums
>   * For a full example please follow the [project setup walkthrough](/lib/project-setup/create-application)

Reviewer 3: ```suggestion
Amplify requires a minimum target platform for iOS (13.0), Android (API level 24), and macOS (10.15). Additional setup is required for some target platforms. Please see the [platform setup](/lib/project-setup/platform-setup) guide for more details on platform specific setup.
```

I had [updated this language](https://github.com/aws-amplify/docs/commit/4b0f4a72f4bb37b31f2e2d6ea05204b244be74ea#diff-731d99bdc314ab25d49f7613b88ab39ae978cbd96644eb5fda62b06da4f1c8db) in a previous PR.

With all the platforms we support now, and the amount of setup required for some of those platforms (macOS in particular) IMO it is too much info to list out all the reqs all over the place. The min platform levels is the most important info, and then customers can view the platform setup guide for all the details.

## Auth

- Changes to API options
- Changes to exception handling
- Changes to nextStep types
- getPlugin for Cognito-specific behavior

> - Changes to Hub events

Reviewer 3: See below for Auth category changes. I was going to leave this as a suggested change, but that doesn't work when there are code snippets.

<details>

- `fetchAuthSession` had several changes. See the example below for how to migrate.
  - `Amplify.Auth.getPlugin()` has been added to return plugin specific sub types. Casting to `CognitoAuthSession` is no longer needed when using this.
  - `getAWSCredentials` has been deprecated. AWS credentials will always be included in the response.
  - `fetchAuthSession` will no longer throw exceptions. Instead, indivdual getters of type `AuthResult` will throw if the underlying operation fails.
    - _Why?_ The previous behavior did not allow for partial successes in offline scenarios when user pool tokens were still valid, but AWS credentials were expired. See [amplify-flutter#760](https://github.com/aws-amplify/amplify-flutter/issues/760#issuecomment-1453559459) for more details.
  - `CognitoAuthSession.credentials`, `CognitoAuthSession.userPoolTokens`, `CognitoAuthSession.`, and ``CognitoAuthSession.` are deprecated and will be removed in a future release.

```dart
// v0
try {
  final session = await Amplify.Auth.fetchAuthSession(
    options: CognitoFetchAuthSessionOptions(getAWSCredentials: true),
  ) as CognitoAuthSession;

  final accessToken = session.userPoolTokens?.accessToken;
  safePrint('accessToken: $accessToken');

  final identityId = session.identityId;
  safePrint('identityId: $identityId');
} on AuthException catch (e) {
  // fetchAuthSession will fail if any of the underlying operations fail.
  safePrint('The operation failed: ${e.message}');
}
```

```dart
// v1
final session = await Amplify.Auth.getPlugin(
  AmplifyAuthCognito.pluginKey,
).fetchAuthSession();

// AuthResult.value will throw if the underlying operation failed.
try {
  final accessToken = session.userPoolTokensResult.value.accessToken.toJson();
  safePrint('accessToken: $accessToken');
} on SignedOutException {
  // the user is not signed in.
} on SessionExpiredException {
  // the users session has expired.
} on NetworkException {
  // the access and/or id token is expired but cannot be refreshed because the
  // user is offline.
}

// `AuthResult.valueOrNull` will return the identityId if present or null,
// but will never throw even if the underlying operation failed.
final identityId = session.identityIdResult.valueOrNull;
safePrint('identityId: $identityId');
```

- The next step returned from auth operations are now enums. See the example below for how to migrate for `Auth.signUp()`.

```dart
// v0
final result = await Amplify.Auth.signUp(/* ... */);
switch (result.nextStep?.signUpStep) {
  case 'CONFIRM_SIGN_UP':
    // prompt user to confirm sign in using nextStep.codeDeliveryDetails
    break;
  case 'DONE':
    // sign up is complete, user can be taken to an authenticated state.
    break;
}

// v1
final result = await Amplify.Auth.signUp(/* ... */);
switch (result.nextStep.signUpStep) {
  case AuthSignUpStep.confirmSignUp:
    // prompt user to confirm sign in using nextStep.codeDeliveryDetails
    break;
  case AuthSignUpStep.done:
    // sign up is complete, user can be taken to an authenticated state.
    break;
}
```

- Auth hub events are now have a specific sub type, `AuthHubEvent` and contain a `type` that is an enum. See the example below for how to migrate.

```dart
// v0
final hubSubscription = Amplify.Hub.listen([HubChannel.Auth], (hubEvent) {
  switch(hubEvent.eventName) {
    case 'SIGNED_IN':
      safePrint('User is signed in.');
      break;
    case 'SIGNED_OUT':
      safePrint('User is signed out.');
      break;
    case 'SESSION_EXPIRED':
      safePrint('The session has expired.');
      break;
    case 'USER_DELETED':
      safePrint('The user has been deleted.');
      break;
  }
});
```

```dart
// v1
final subscription = Amplify.Hub.listen(HubChannel.Auth, (AuthHubEvent event) {
  switch (event.type) {
    case AuthHubEventType.signedIn:
      safePrint('User is signed in.');
      break;
    case AuthHubEventType.signedOut:
      safePrint('User is signed out.');
      break;
    case AuthHubEventType.sessionExpired:
      safePrint('The session has expired.');
      break;
    case AuthHubEventType.userDeleted:
      safePrint('The user has been deleted.');
      break;
  }
});
```

- Cognito specific API options such as `CognitoSignInOptions` are deprecated in favor of `pluginOptions`. The classes will be removed in a future release. See below for an example of how to migrate for `Auth.signIn()`.

```dart
// v0
final result = await Amplify.Auth.signIn(
  username: username,
  password: password,
  options: const CognitoSignInOptions(
    clientMetadata: {'data': 'value'},
  ),
);

// v1
final result = await Amplify.Auth.signIn(
  username: username,
  password: password,
  options: const SignInOptions(
    pluginOptions: CognitoSignInPluginOptions(
      clientMetadata: {'data': 'value'},
    ),
  ),
);
```
</details>

Author: Just confirming these are the changes you already comitted?

## Analytics

- `AnalyticsProperties` renamed to `CustomProperties` 
- `AnalyticsUserProfile` renamed to `UserProfile`
- `AnalyticsUserProfileLocation` renamed to `UserProfileLocation`

> - In v1 you can now specify autoFlushEventsInterval in seconds in amplifyConfiguration.dart to change how often cached events are flushed. v0 does not support this due to differences in iOS and Android implementations.

Reviewer 3: @fjnoyp - This is phrased as a new feature. If the goal here is to note that `autoFlushEventsInterval` is now in seconds, whereas it was is ms on some platforms, let's re-phrase this to call that out.

Reviewer 2: In v0, the `autoFlushEventsInterval` in `amplifyconfiguration.dart` was read as seconds on iOS and milliseconds on Android.  v1 fixes this behavior by reading `autoFlushEventsInterval` as seconds across all platforms.

> - Previously Analytics was built on top of existing Amplify Android and iOS functionality. Those platforms behaved differently but now behave the same:
>     - in v1 flushing occurs at 30 second intervals. In v0 default auto flush iOS was 60 seconds, Android was 30 seconds.
>     - in v1 session pause and resume are not automatically tracked. In v0 iOS auto tracked session pause and resume events. 

Reviewer 3: @fjnoyp - I would focus on the behavior change, not why it use to behave a different way.

```suggestion
  - in v1 flushing occurs at 30 second intervals. In v0 default auto flush iOS was 60 seconds, Android was 30 seconds.
  - in v1 session pause and resume are not automatically tracked. In v0 iOS auto tracked session pause and resume events. 
```

- You can now save to the `UserAttributes` field of a Pinpoint Endpoint by sending a `AWSPinpointUserProfile` to `identifyUser`.

> - Analytics Endpoint metadata is stored in a different namespace/location.  Existing metadata will automatically be migrated.

Reviewer 3: @fjnoyp - I would omit this. It is not breaking and requires no action from the user.
```suggestion
```

- Analytics cached event data is now stored differently. Existing cached analytics events will not be migrated.

> - There are now typed `AnalyticsException`s for specific exception cases. 

Reviewer 3: ```suggestion
- There are now typed `AnalyticsException`s for specific exception cases. See [the documentation on Pub](https://pub.dev/documentation/amplify_analytics_pinpoint/latest/amplify_analytics_pinpoint/amplify_analytics_pinpoint-library.html) for a full list of exceptions.
```

Lets link to the docs on pub. It clearly shows the full list. That link directs to the latest stable docs, which right now is 0.6.13, but will be 1.0 after we publish.

Note: I am not sure if that link is formatted correctly for the docs site.

Author: link format worked locally for me, update coming
